SourceKitStressTester/Sources/StressTester/SourceKitDocument.swift

```diff
@@ -42,29 +41,42 @@ struct SourceKitDocument {
     let needsLineColumn: Bool = true
   }
 
-  init(_ file: String, args: [String], connection: SourceKitdService, containsErrors: Bool = false) {
-    self.file = file
+  init(args: CompilerArgs,
+       connection: SourceKitdService,
+       containsErrors: Bool = false) {
     self.args = args
-    self.containsErrors = containsErrors
     self.connection = connection
+    self.containsErrors = containsErrors
     self.diagEngine.addConsumer(EmptyDiagConsumer())
   }
 
-  mutating func open(state: SourceState? = nil) throws -> (SourceFileSyntax, SourceKitdResponse) {
+  mutating func open(rewriteMode: RewriteMode) throws -> (SourceFileSyntax, SourceKitdResponse) {
+    let source = try! String(contentsOf: args.forFile, encoding: .utf8)
+    sourceState = SourceState(rewriteMode: rewriteMode, content: source)
+    return try openOrUpdate(path: args.forFile.path)
+  }
+
+  mutating func update(updateSource: (inout SourceState) -> Void) throws -> (SourceFileSyntax, SourceKitdResponse) {
+    var sourceState = self.sourceState!
+    try close()
+    updateSource(&sourceState)
+    self.sourceState = sourceState
+    return try openOrUpdate(source: sourceState.source)
+  }
+
+  private mutating func openOrUpdate(path: String? = nil, source: String? = nil)
+  throws -> (SourceFileSyntax, SourceKitdResponse) {
     let request = SourceKitdRequest(uid: .request_EditorOpen)
-    if let state = state {
-      sourceState = state
-      request.addParameter(.key_SourceText, value: state.source)
-    } else {
-      request.addParameter(.key_SourceFile, value: file)
+    if let path = path {
+      request.addParameter(.key_SourceFile, value: path)
+    } else if let sourceState = sourceState {
+      request.addParameter(.key_SourceText, value: sourceState.source)
     }
```

Comment:
should this use the 'source' param instead?

Reply:
Nice catch, though I suppose it didn't actually matter since it was being updated before the call. Maybe I should also set it in this method?

Reply:
Then again, it's only set in open as well. I'll just fix up sourceState -> source